src/libasr/pass/pass_utils.h

```diff
@@ -314,9 +319,38 @@ namespace LCompilers {
                 }
 
                 void visit_FunctionCall(const ASR::FunctionCall_t& x) {
-                    if( fill_function_dependencies ) {
-                        function_dependencies.push_back(al, ASRUtils::symbol_name(x.m_name));
+                    if( fill_function_dependencies &&
+                        ASRUtils::symbol_symtab(x.m_name) != nullptr &&
+                        ASRUtils::symbol_parent_symtab(x.m_name)->get_counter() !=
+                        current_scope->get_counter()) {
+                        if (ASR::is_a<ASR::Function_t>(*x.m_name)) {
+                            ASR::Function_t *f = ASR::down_cast2<ASR::Function_t>(current_scope->asr_owner);
+
+                            // Check is x.m_name is not an argument.
+                            bool is_arg = false;
+
+                            for (size_t i=0; i<f->n_args; i++) {
+                                ASR::Var_t *arg = ASR::down_cast<ASR::Var_t>(f->m_args[i]);
+
+                                if (arg->m_v == x.m_name) {
+                                    is_arg = true;
+                                    break;
+                                }
+                            }
+
+                            if (!is_arg) {
+                                function_dependencies.push_back(al, ASRUtils::symbol_name(x.m_name));
+                            }
+                        }
                     }
+
+                    if (fill_function_dependencies && ASR::is_a<ASR::ExternalSymbol_t>(*x.m_name)) {
+                        ASR::ExternalSymbol_t* external_symbol = ASR::down_cast<ASR::ExternalSymbol_t>(x.m_name);
+                        if (!ASRUtils::is_present_in_current_scope(external_symbol, current_scope)) {
+                            function_dependencies.push_back(al, ASRUtils::symbol_name(x.m_name));
+                        }
+                    }
+
```

Comment:
These two separate `if`s can be combined. Right now its hard to read through. We discussed this as well in the meeting.